hail/src/main/scala/is/hail/expr/ir/Infer.scala

```diff
@@ -1,89 +1,101 @@
       case GroupByKey(collection) =>
-        val elt = coerce[TBaseStruct](coerce[TArray](collection.typ).elementType)
-        TDict(elt.types(0), TArray(elt.types(1)))
+        val elt = coerce[PBaseStruct](coerce[PArray](collection.pType).elementType)
+        // FIXME requiredness
+        PDict(elt.types(0), PArray(elt.types(1), required = false), collection.pType.required)
       case ArrayMap(a, name, body) =>
-        TArray(-body.typ, a.typ.required)
+        // FIXME: requiredness artifact of former IR bug, remove when possible
+        PArray(body.pType.setRequired(false), a.typ.required)
       case ArrayFilter(a, name, cond) =>
-        a.typ
+        PArray(coerce[PArray](a.pType).elementType, a.pType.required)
       case ArrayFlatMap(a, name, body) =>
-        TArray(coerce[TContainer](body.typ).elementType)
+        PArray(coerce[PContainer](body.pType).elementType, a.pType.required)
       case ArrayFold(a, zero, accumName, valueName, body) =>
-        assert(body.typ == zero.typ)
-        zero.typ
+        assert(body.pType == zero.pType)
+        zero.pType
       case ArrayScan(a, zero, accumName, valueName, body) =>
-        assert(body.typ == zero.typ)
-        TArray(zero.typ)
+        assert(body.pType == zero.pType)
+        PArray(zero.pType)
       case AggFilter(_, aggIR) =>
-        aggIR.typ
+        aggIR.pType
       case AggExplode(array, name, aggBody) =>
-        aggBody.typ
+        aggBody.pType
       case AggGroupBy(key, aggIR) =>
-        TDict(key.typ, aggIR.typ)
+        PDict(PType.canonical(key.pType), aggIR.pType)
       case ApplyAggOp(_, _, _, aggSig) =>
-        AggOp.getType(aggSig)
+        PType.canonical(AggOp.getType(aggSig))
       case ApplyScanOp(_, _, _, aggSig) =>
-        AggOp.getType(aggSig)
+        PType.canonical(AggOp.getType(aggSig))
       case MakeStruct(fields) =>
-        TStruct(fields.map { case (name, a) =>
-          (name, a.typ)
+        // FIXME requiredness
+        PStruct(required = false,
+          fields.map { case (name, a) =>
+          (name, a.pType)
         }: _*)
       case SelectFields(old, fields) =>
-        val tbs = coerce[TStruct](old.typ)
-        TStruct(fields.map { id => (id, tbs.field(id).typ) }: _*)
+        val tbs = coerce[PStruct](old.pType)
+        tbs.selectFields(fields)
       case InsertFields(old, fields) =>
-        old.typ.asInstanceOf[TStruct].insertFields(fields.map(x => (x._1, x._2.typ)))
+        val tbs = coerce[PStruct](old.pType)
+        tbs.insertFields(fields.map(f => (f._1, f._2.pType)))
       case GetField(o, name) =>
-        val t = coerce[TStruct](o.typ)
+        val t = coerce[PStruct](o.pType)
         assert(t.index(name).nonEmpty, s"$name not in $t")
-        -t.field(name).typ
-      case MakeTuple(types) =>
-        TTuple(types.map(_.typ): _*)
+        val fd = t.field(name).typ
+        fd.setRequired(t.required && fd.required)
+      case MakeTuple(values) =>
+        // FIXME requiredness
```

Comment:
?

Reply:
The tuple should be required=True when possible.